browser/brave_ads/search_ad_metadata_handler.cc

```diff
@@ -0,0 +1,306 @@
+        }
+        found_attributes.insert(*it);
+
+        if (!SetSearchAdProperty(ad_property, &search_ad)) {
+          LOG(ERROR) << "Cannot read search ad attribute value: "
+                     << ad_property->name;
+          return SearchResultAdsList();
+        }
+      }
+
+      // Not all of attributes were specified.
+      if (found_attributes.size() != kSearchAdAttributes.size()) {
+        std::vector<base::StringPiece> absent_attributes;
+        std::set_difference(kSearchAdAttributes.begin(),
+                            kSearchAdAttributes.end(), found_attributes.begin(),
+                            found_attributes.end(),
+                            std::back_inserter(absent_attributes));
+
+        LOG(ERROR) << "Some of search ad attributes were not specified: "
+                   << base::JoinString(absent_attributes, ", ");
+
+        return SearchResultAdsList();
+      }
+
+      search_ads_list.push_back(std::move(search_ad));
+    }
+
+    // Creatives has been parsed.
+    break;
+  }
+
+  return search_ads_list;
+}
+
+SearchResultAdsList ParseMetadataEntities(blink::mojom::WebPagePtr web_page) {
+  for (const auto& entity : web_page->entities) {
+    if (entity->type != kSearchAdLinkDataType) {
+      continue;
+    }
+
+    absl::optional<SearchResultAdsList> search_ads_list =
+        ParseEntityProperties(entity);
+
+    if (search_ads_list) {
+      return std::move(*search_ads_list);
+    }
+  }
+
+  return SearchResultAdsList();
+}
+
+}  // namespace
+
+SearchAdMetadataHandler::SearchAdMetadataHandler() = default;
+
+SearchAdMetadataHandler::~SearchAdMetadataHandler() = default;
+
+bool SearchAdMetadataHandler::IsAllowedBraveSearchHost(const GURL& url) const {
+  if (!url.is_valid() || !url.SchemeIs(url::kHttpsScheme)) {
+    return false;
+  }
+  const std::string host = url.host();
+  return base::Contains(kVettedBraveSearchHosts, host);
+}
+
+void SearchAdMetadataHandler::RetrieveSearchAdMetadata(
+    content::RenderFrameHost* render_frame_host,
+    RetrieveSearchAdMetadataCallback callback) {
+  DCHECK(render_frame_host);
+  DCHECK(IsAllowedBraveSearchHost(render_frame_host->GetLastCommittedURL()));
+  DCHECK(!document_metadata_.is_bound());
+
+  render_frame_host->GetRemoteInterfaces()->GetInterface(
+      document_metadata_.BindNewPipeAndPassReceiver());
+  DCHECK(document_metadata_.is_bound());
+
+  document_metadata_->GetEntities(base::BindOnce(
+      &SearchAdMetadataHandler::OnRetrieveSearchAdMetadataEntities,
+      weak_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void SearchAdMetadataHandler::OnRetrieveSearchAdMetadataEntities(
+    RetrieveSearchAdMetadataCallback callback,
+    blink::mojom::WebPagePtr web_page) {
+  document_metadata_.reset();
+
+  if (!web_page) {
+    std::move(callback).Run(SearchResultAdsList());
+    return;
+  }
+
+  // TODO(https://github.com/brave/brave-browser/issues/20852):
+  // Send search ads list to Ads Service.
```

Comment:
@aseren do we need to send search ads list to ads service both here and in the ads tab helper?

Reply:
`mojom::SearchResultAd` now contains the `placement_id` (see spec)

Reply:
I've forgotten to remove this comment. The right place is `AdsTabHelper::OnRetrieveSearchAdMetadata`